runtime/plaid/src/cache/redis/mod.rs

```diff
@@ -0,0 +1,184 @@
+            "@".to_string()
+        };
+
+        format!("redis://{username}{password}{at}{hostname}{port}")
+    }
+}
+
+impl RedisCache {
+    pub async fn new(config: Config, max_capacity: LimitedAmount) -> Result<Self, CacheError> {
+        let conn_string = config.build_connection_string();
+        let client = redis::Client::open(conn_string).map_err(|e| {
+            CacheError::CacheInitError(format!("Could not create redis client: {e}"))
+        })?;
+        let connection_manager = client
+            .get_connection_manager()
+            .await
+            .map_err(|e| CacheError::CacheAccessError(e.to_string()))?;
+
+        let eviction_policy = config.eviction_policy.unwrap_or_default();
+        Ok(Self {
+            connection_manager,
+            eviction_policy,
+            max_capacity,
+        })
+    }
+}
+
+#[async_trait]
+impl super::CacheProvider for RedisCache {
+    async fn put(
+        &self,
+        namespace: &str,
+        key: &str,
+        value: &str,
+    ) -> Result<Option<String>, CacheError> {
+        let mut connection = self.connection_manager.clone();
+
+        // Check if we need to evict an entry. If so, do it to make space for the upcoming insertion.
+        // Note - We do this in an "optimistic" way, without locking anything.
+        let old = match self.eviction_policy {
+            EvictionPolicy::NoEviction => {
+                let (old, ()): (Option<String>, ()) = redis::pipe()
+                    .cmd("HGET")
+                    .arg(namespace)
+                    .arg(key)
+                    .cmd("HSET")
+                    .arg(namespace)
+                    .arg(key)
+                    .arg(value)
+                    .query_async(&mut connection)
+                    .await
+                    .map_err(|e| CacheError::PutError(e.to_string()))?;
+
+                old
+            }
+            EvictionPolicy::RandomEviction => {
+                // Figure out how many entries can be stored in this cache.
+                // If we have a value for this namespace (i.e., this module), use
+                // that; otherwise fallback on the default value.
+                let max_entries = self
+                    .max_capacity
+                    .module_overrides
+                    .get(namespace)
+                    .map(|v| *v)
+                    .unwrap_or(self.max_capacity.default);
+
+                let script = redis::Script::new(include_str!("./eviction.redis"));
+
+                let (old, evicted): (Option<String>, bool) = script
+                    .key(namespace)
+                    .arg(key)
+                    .arg(value)
+                    .arg(max_entries)
+                    .invoke_async(&mut connection)
+                    .await
+                    .map_err(|e| CacheError::PutError(e.to_string()))?;
+
+                if evicted {
+                    debug!("An item has been evicted from the Redis cache");
+                }
```

Comment:
I can see this being useful but I'm wondering if this is something that should be interrogated on the Redis side? Especially if collecting the information on if a key was evicted costs extra compute time